File: node/m10-sdk/src/types/action.ts
import { google, m10 } from "../../protobufs";
import { M10Error } from "../error";
import type { TxId } from "../ids";
import { AccountId } from "../ids/account_id";
import { isSome, unwrap } from "../utils";



export class Target {
    public accountId: Option<AccountId>;


    public constructor(accountId?: Option<AccountId>) {
        this.accountId = accountId;
    }

    public into(): m10.sdk.transaction.Target {
        return (
            isSome(this.accountId)
                ? new m10.sdk.transaction.Target({ accountId: this.accountId.toUint8Array() })
                : new m10.sdk.transaction.Target({ anyAccount: new google.protobuf.Empty() })
        );
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     * @throws  {M10Error.InvalidAccountId}     if invalid account id was used
     */
    public static tryFrom(target: m10.sdk.transaction.ITarget): Target {
        return (
            isSome(target.anyAccount)
                ? new Target()
                : new Target(
                    AccountId.fromUint8Array(
                        unwrap(target.accountId, M10Error.InvalidTransaction()),
                    ),
                )
        );
    }
}

export interface IAction {
    txId: TxId;
    contextId: Uint8Array;
    name: string;
    target: Target;
    payload: Uint8Array;
}

export class Action implements IAction {
    public txId: TxId;
    public contextId: Uint8Array;
    public name: string;
    public target: Target;
    public payload: Uint8Array;

    public constructor(properties: IAction) {
        this.txId = properties.txId;
        this.contextId = properties.contextId;
        this.name = properties.name;
        this.target = properties.target;
        this.payload = properties.payload;
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     * @throws  {M10Error.InvalidAccountId}     if invalid account id was used
     */
    public static tryFrom(action: m10.sdk.transaction.IAction): Action {
        return new Action({
            txId: unwrap(action.txId, M10Error.InvalidTransaction()),
            contextId: unwrap(action.contextId, M10Error.InvalidTransaction()),
            name: unwrap(action.name, M10Error.InvalidTransaction()),
            target: Target.tryFrom(unwrap(action.target, M10Error.InvalidTransaction())),
            payload: unwrap(action.payload, M10Error.InvalidTransaction()),
        });
    }

    /**
     * @throws  {M10Error.InvalidTransaction}   if any of the required fields are missing
     */
    public static tryFromFinalizedTransaction(txn: m10.sdk.IFinalizedTransaction): Action {
        return new Action({
            txId: unwrap(txn.response?.txId, M10Error.InvalidTransaction()),
            contextId: unwrap(txn.request?.contextId, M10Error.InvalidTransaction()),
            name: unwrap(txn.request?.data?.invokeAction?.name, M10Error.InvalidTransaction()),
            target: new Target(),
            payload: unwrap(txn.request?.data?.invokeAction?.payload, M10Error.InvalidTransaction()),
        });
    }
}

File: rust/sdk/src/types/action.rs
use crate::account::{AccountId, AccountIdError};
use crate::error::M10Error;
use crate::types::TxId;
use crate::TransactionExt;
use core::convert::{From, TryFrom};
use core::result::Result;
use m10_protos::sdk;
use m10_protos::sdk::transaction_data::Data;
use serde::Serialize;
use serde_with::serde_as;

#[cfg_attr(feature = "format", derive(parse_display::Display))]
#[derive(Clone, Debug, Serialize)]
pub enum Target {
    Any,
    #[cfg_attr(feature = "format", display("Account({0})"))]
    Account(AccountId),
}

impl From<Target> for sdk::Target {
    fn from(target: Target) -> Self {
        let target = match target {
            Target::Any => sdk::target::Target::AnyAccount(()),
            Target::Account(id) => sdk::target::Target::AccountId(id.to_vec()),
        };
        Self {
            target: Some(target),
        }
    }
}

impl TryFrom<sdk::Target> for Target {
    type Error = M10Error;

    fn try_from(target: sdk::Target) -> Result<Self, Self::Error> {
        Ok(match target.target {
            Some(sdk::target::Target::AnyAccount(())) => Target::Any,
            Some(sdk::target::Target::AccountId(id)) => {
                Target::Account(AccountId::try_from_be_slice(&id)?)
            }
            None => return Err(M10Error::InvalidAccountId(AccountIdError::MissingRoot)),
        })
    }
}

#[serde_as]
#[derive(Clone, Debug, Serialize)]
pub struct Action {
    pub tx_id: TxId,
    #[serde_as(as = "serde_with::hex::Hex")]
    pub context_id: Vec<u8>,
    pub name: String,
    pub from_account: AccountId,
    pub target: Target,
    pub payload: Vec<u8>,
    // @sadroeck TODO: Add timestamps to protobuf
    // pub timestamp: SystemTime,
}

#[cfg(feature = "format")]
impl std::fmt::Display for Action {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            tx_id,
            context_id,
            name,
            target,
            payload,
            ..
        } = self;
        write!(
            f,
            "Action{{ tx_id={tx_id} name={name} context={} target={target} payload={}",
            hex::encode(context_id),
            hex::encode(payload)
        )
    }
}

impl TryFrom<sdk::Action> for Action {
    type Error = M10Error;

    fn try_from(action: sdk::Action) -> Result<Self, Self::Error> {
        Ok(Self {
            tx_id: action.tx_id,
            context_id: action.context_id,
            // timestamp: UNIX_EPOCH + Duration::from_micros(action.timestamp),
            name: action.name,
            from_account: AccountId::try_from_be_slice(&action.from_account)?,
            target: Target::try_from(action.target.unwrap())?,
            payload: action.payload,
        })
    }
}

impl TryFrom<sdk::FinalizedTransaction> for Action {
    type Error = M10Error;

    fn try_from(txn: sdk::FinalizedTransaction) -> Result<Self, Self::Error> {
        let response = txn.response.as_ref().ok_or(M10Error::InvalidTransaction)?;
        let context_id = txn
            .request
            .as_ref()
            .ok_or(M10Error::InvalidTransaction)?
            .context_id
            .clone();
        let tx_id = response.tx_id;
        // let timestamp = UNIX_EPOCH + Duration::from_micros(response.timestamp);
        match txn.into_data().ok_or(M10Error::InvalidTransaction)? {
            Data::InvokeAction(action) => Ok(Self {
                tx_id,
                context_id,
                // timestamp,
                name: action.name,
                from_account: AccountId::try_from_be_slice(&action.from_account)?,
                target: Target::Any,
                payload: action.payload,
            }),
            _ => Err(M10Error::InvalidTransaction),
        }
    }
}
